### ISSUE TITLE
Angular 17 Issues

### ISSUE THREAD

**Issue by A065639** [ng-packagr/ng-packagr#2722]
## Type of Issue

```
[ X ] Bug Report
[ ] Feature Request
```

## Description

Unable to complete a containerized AzDO pipeline build after the Angular 17 and ng-packagr version 17 upgrade. 

#### How To Reproduce

We have a Angular repo that is a library we publish to our internal package hosting library. We use ng-packagr to help accomplish this task. After the recent upgrade, we are able to build locally without issue. We can even create and install the library locally. However, when we attempt to run a CI/CD build we receive failures (shown below). 

We build our package locally using a Windows machine, the missing package is specific to a Linux environment so we cannot install it locally. We can force it to be installed during our CI/CD pipeline in the DOCKER file but this feels like a hacky way to accomplish this task. 

Curious if others have had this issue or if we are doing something wrong. 

`Cannot find module '@rollup/rollup-linux-x64-gnu'
Require stack:
- /opt/app-root/src/node_modules/ng-packagr/node_modules/rollup/dist/native.js
- /opt/app-root/src/node_modules/ng-packagr/node_modules/rollup/dist/shared/parseAst.js
- /opt/app-root/src/node_modules/ng-packagr/node_modules/rollup/dist/shared/rollup.js
- /opt/app-root/src/node_modules/ng-packagr/node_modules/rollup/dist/rollup.js
- /opt/app-root/src/node_modules/ng-packagr/lib/flatten/rollup.js
- /opt/app-root/src/node_modules/ng-packagr/lib/ng-package/entry-point/write-bundles.transform.js
- /opt/app-root/src/node_modules/ng-packagr/lib/ng-package/entry-point/write-bundles.di.js
- /opt/app-root/src/node_modules/ng-packagr/lib/ng-package/entry-point/entry-point.di.js
- /opt/app-root/src/node_modules/ng-packagr/lib/packagr.js
- /opt/app-root/src/node_modules/ng-packagr/lib/commands/build.command.js
- /opt/app-root/src/node_modules/ng-packagr/public_api.js
- /opt/app-root/src/node_modules/@angular-devkit/build-angular/src/builders/ng-packagr/index.js
- /opt/app-root/src/node_modules/@angular-devkit/architect/node/node-modules-architect-host.js
- /opt/app-root/src/node_modules/@angular-devkit/architect/node/index.js
- /opt/app-root/src/node_modules/@angular/cli/src/command-builder/architect-base-command-module.js
- /opt/app-root/src/node_modules/@angular/cli/src/command-builder/architect-command-module.js
- /opt/app-root/src/node_modules/@angular/cli/src/commands/build/cli.js
- /opt/app-root/src/node_modules/@angular/cli/src/commands/command-config.js
- /opt/app-root/src/node_modules/@angular/cli/src/command-builder/command-runner.js
- /opt/app-root/src/node_modules/@angular/cli/lib/cli/index.js
- /opt/app-root/src/node_modules/@angular/cli/lib/init.js
`

#### Version Information

```
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 17.0.0
Node: 18.18.2
Package Manager: npm 9.8.1
OS: linux x64

Angular: 17.0.2
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1700.0
@angular-devkit/build-angular   17.0.0
@angular-devkit/core            17.0.0
@angular-devkit/schematics      17.0.0
@angular/cdk                    17.0.0
@angular/cli                    17.0.0
@angular/material               17.0.0
@schematics/angular             17.0.0
ng-packagr                      17.0.0
rxjs                            7.8.1
typescript                      5.2.2
zone.js                         0.14.2


_Please include all version numbers that might be relevant, e.g. third-party libraries_

**Comment by alan-agius4** [comment]
That is strange as the mentioned package is published on NPM (https://www.npmjs.com/package/@rollup/rollup-linux-x64-gnu)

Are you consuming packages directly from the public NPM registry or you have a proxy or maybe caching node_modules that might be out of sync?

**Comment by JoostK** [comment]
I suspect what is happening is that you're perhaps installing from Windows, then copying to inside a Docker container. That is a problem, since various dependencies download native binaries depending on the platform they are installed on.

Some package managers allow to download optional dependencies for multiple platforms (recently released in a `pnpm` update, I don't know about support in other package managers), but that may still pose problems for packages that use postinstall scripts to download native binaries itself, like the `node-gyp`   ecosystem.

**Comment by A065639** [comment]
> That is strange as the mentioned package is published on NPM (https://www.npmjs.com/package/@rollup/rollup-linux-x64-gnu)
> 
> Are you consuming packages directly from the public NPM registry or you have a proxy or maybe caching node_modules that might be out of sync?

We are using a proxy but I verified the package is in the registry and can be installed directly within a Linux container.

**Comment by A065639** [comment]
> I suspect what is happening is that you're perhaps installing from Windows, then copying to inside a Docker container. That is a problem, since various dependencies download native binaries depending on the platform they are installed on.
> 
> Some package managers allow to download optional dependencies for multiple platforms (recently released in a `pnpm` update, I don't know about support in other package managers), but that may still pose problems for packages that use postinstall scripts to download native binaries itself, like the `node-gyp` ecosystem.

This is what I suspect as well, not sure how to get around this one at the moment as we will always build our applications using Windows machines then build and deploy them using Linux containers. To my knowledge, this is the first time we have encountered such an issue.

**Comment by alan-agius4** [comment]
Is the npm installation being done on the Linux container or node_modules are being copied over from Windows?

**Comment by A065639** [comment]
> Is the npm installation being done on the Linux container or node_modules are being copied over from Windows?

In the Linux container we were originally calling "npm ci" which would install what is in our package-lock.json created on a Windows machine. 

I tried removing the package-lock.json and running "npm i" to install packages and create the package-lock.json in the Linux container but I got the same result. We are digging into the logging here to see what was installed during this build.

**Comment by A065639** [comment]
We continue to dig into this one but with no luck. We suspect that our developers are no longer able to share a package-lock.json file across Windows/Linux environments which presents challenges. 

It feels like this reduces the compatibility of the ng-packagr package, is there a better way to integrate rollup into this package? rollup-wasm? 

Is this a bug in the latest rollup migration from 3 > 4?

**Comment by alan-agius4** [comment]
Yeah, this does appear to be a problem caused by Rollup version 4. Using `@rollup/wasm-node` should solve the problem.

**Comment by A065639** [comment]
Is this something the group could implement within ng-packgr?

**Comment by alan-agius4** [comment]
I have a PR in flight to fix this. https://github.com/ng-packagr/ng-packagr/pull/2725

**Comment by A065639** [comment]
Fantastic! I look forward to testing it out!

**Comment by alan-agius4** [comment]
Released as 17.0.2

**Comment by A065639** [comment]
Fantastic! I was able to pull it in and build. We will continue verification but for now, this is progress and considered resolved.
